src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices
+{
+    [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+    internal sealed partial class TypeMapLazyDictionary
+    {
+        private readonly LazyExternalTypeDictionary _externalTypeMap;
+        private readonly LazyProxyTypeDictionary _proxyTypeMap;
+
+        private ref struct CallbackContext
+        {
+            [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+            public CallbackContext()
+            {
+                ExternalTypeMap = new LazyExternalTypeDictionary();
+                ProxyTypeMap = new LazyProxyTypeDictionary();
+            }
+
+            public LazyExternalTypeDictionary ExternalTypeMap;
+            public LazyProxyTypeDictionary ProxyTypeMap;
+        }
+
+        // See assemblynative.hpp for native version.
+        public unsafe struct ProcessAttributesCallbackArg
+        {
+            public void* Utf8String1;
+            public void* Utf8String2;
+            public void* Utf8String3;
+            public int StringLen1;
+            public int StringLen2;
+            public int StringLen3;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypeMapLazyDictionary_ProcessAttributes")]
+        private static unsafe partial void ProcessAttributes(
+            QCallAssembly assembly,
+            QCallTypeHandle groupType,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newExternalTypeEntry,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
+            CallbackContext* context);
+
+        private static TypeName CreateNewTypeName(string typeName, string assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = new TypeName(typeName, asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
```

Comment:
This can use arraypool so reduce allocation profile of the whole operation. (`TypeNameParser.Parse` takes Span.)

Reply:
Done.